digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataExporter/AssetReportCsvExporter.scala

```diff
@@ -0,0 +1,264 @@
+      }
+    }
+
+    def extractData(municipality: Int): List[AssetReport] = {
+      if (withTransaction) {
+        withDynTransaction {
+          executeQuery(municipality)
+        }
+      }
+      else {
+        executeQuery(municipality)
+      }
+    }
+
+    def getUserAndMunicipalitiesInfo(): (Seq[User], List[MunicipalityInfo]) = {
+      if(withTransaction) {
+        withDynSession {
+          val userResults = userProvider.getUsers()
+          val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+          (userResults, municResults)
+        }
+      }
+      else{
+        val userResults = userProvider.getUsers()
+        val municResults = municipalityDao.getMunicipalitiesNameAndIdByCode(municipalitiesList.toSet)
+
+        (userResults, municResults)
+      }
+    }
+
+
+    val (users, municipalitiesInfo) = getUserAndMunicipalitiesInfo()
+
+    val operatorUsers = users.filter(_.isOperator())
+    val operatorUsersUsername = operatorUsers.map(_.username)
+
+    municipalitiesList.flatMap{ municipality =>
+
+      val municipalityUsers = users.filter( _.isAuthorizedToWrite(municipality) )
+                                    .filterNot( user => operatorUsersUsername.contains(user.username) )
+                                    .map( _.username )
+
+      val linearAssets = extractData(municipality)
+
+      if ( linearAssets != null && linearAssets.nonEmpty) {
+        val linearAssetsGroupedBy = linearAssets.groupBy(_.assetType)
+
+        assetTypesList.flatMap { assetType =>
+          if (linearAssetsGroupedBy.contains(assetType)) {
+            val currentAssetType = linearAssetsGroupedBy(assetType)
+
+            val (operatorUser, operatorDate) = extractUserAndDateTime(currentAssetType, operatorUsersUsername)
+            val (municipalityUser, municipalityDate) = extractUserAndDateTime(currentAssetType, municipalityUsers)
+
+            val firstElem = currentAssetType.head
+
+            Some( ExportAssetReport(municipalitiesInfo.filter(_.id == municipality).head.name, firstElem.assetType, currentAssetType.size,
+                              firstElem.assetNameFI, firstElem.assetGeometryType, operatorUser, operatorDate, municipalityUser, municipalityDate) )
+          }
+          else
+            None
+        }
+
+      }
+      else
+        None
+    }
+  }
+
+
+  def extractDateToString(dateTime: Option[DateTime]): String = {
+    dateTime match {
+      case Some(date) =>
+        DateParser.dateToString(date, DateParser.DatePropertyFormat)
+      case _ => ""
+    }
+  }
+
+
+  def extractRecords( exportAssetReports: List[ExportAssetReport], withTransaction: Boolean = true): String = {
+
+    def getTotalNewLawTrafficSigns(municipalityName: String): Int = {
+      if(withTransaction) {
+        withDynSession {
+          val municipalityInfo = municipalityDao.getMunicipalityIdByName(municipalityName)
```

Comment:
This is an old method.
And it is being called when the users request the traffic sign information (and not to all assets)

Reply:
This method returns a list because a municipality name is not specific unique. In the moment it is, but there is no rule that municipalities shouldn't have equal names.

Reply:
Changed the getMunicipalityIdByName to another name and returns now a option